csharp/src/Google.Protobuf/Collections/MapField.cs

```diff
@@ -221,17 +227,15 @@ public bool TryGetValue(TKey key, out TValue value)
             }
         }
 
-        // TODO: Make these views?
-
         /// <summary>
         /// Gets a collection containing the keys in the map.
         /// </summary>
-        public ICollection<TKey> Keys { get { return list.Select(t => t.Key).ToList(); } }
+        public ICollection<TKey> Keys { get { return new MapView<TKey>(this, pair => pair.Key, ContainsKey); } }
```

Comment:
Mmm. It's certainly not hard to do, but I'm not sure how much benefit there is - at the cost of a couple of extra fields. Once we eventually fix the earlier TODO and don't allocate a map/list until we need one, map fields should be very cheap in terms of memory.

I guess two fields (16 bytes) isn't a terribly heavy price - it depends on how many maps we end up seeing.